File: apps/api/src/domain/repository.ts
import type {
  CyclingEvent,
  Registration,
  RegistrationResult,
  Roadbook,
  UserRegistrationItem,
} from "./models.js";

export interface EventListQuery {
  status?: CyclingEvent["status"];
  difficulty?: CyclingEvent["difficulty"];
  cursor?: string;
  limit: number;
}

export interface Page<T> {
  items: T[];
  nextCursor: string | null;
}

export interface RegisterCommand {
  eventId: string;
  userId: string;
  idempotencyKey: string;
  abilityConfirmed: boolean;
  equipmentConfirmed: boolean;
  waiverVersion: string;
  phoneEncrypted: string;
  emergencyContactEncrypted: string;
  bikeType: string;
  now: Date;
}

export interface Repository {
  createEvent(event: CyclingEvent): Promise<CyclingEvent>;
  updateEvent(event: CyclingEvent): Promise<CyclingEvent>;
  getEvent(id: string): Promise<CyclingEvent | null>;
  listEvents(query: EventListQuery): Promise<Page<CyclingEvent>>;
  createRoadbook(roadbook: Roadbook): Promise<Roadbook>;
  getRoadbook(id: string): Promise<Roadbook | null>;
  listRoadbooks(limit: number, cursor?: string): Promise<Page<Roadbook>>;
  registerAtomically(command: RegisterCommand): Promise<RegistrationResult>;
  cancelRegistrationAtomically(eventId: string, userId: string, now: Date): Promise<RegistrationResult>;
  getRegistration(eventId: string, userId: string): Promise<Registration | null>;
  listRegistrationsByUser(userId: string): Promise<UserRegistrationItem[]>;
}

File: apps/api-go/internal/domain/repository.go
package domain

import (
	"context"
	"time"
)

type EventListQuery struct {
	Status     *EventStatus
	Difficulty *Difficulty
	Cursor     *string
	Limit      int
}

type RegisterCommand struct {
	EventID                   string
	UserID                    string
	IdempotencyKey            string
	AbilityConfirmed          bool
	EquipmentConfirmed        bool
	WaiverVersion             string
	PhoneEncrypted            string
	EmergencyContactEncrypted string
	BikeType                  string
	Now                       time.Time
}

type Repository interface {
	GetUserProfile(context.Context, string) (*UserProfile, error)
	UpsertUserProfile(context.Context, UserProfile) (UserProfile, error)
	GetUserIDByPhoneHash(context.Context, string) (*string, error)
	BindUserPhone(context.Context, string, string, string, string, time.Time) error
	CreateEvent(context.Context, Event) (Event, error)
	UpdateEvent(context.Context, Event) (Event, error)
	UpdateEventWithChange(context.Context, Event, EventChange) (Event, error)
	GetEvent(context.Context, string) (*Event, error)
	ListEventChanges(context.Context, string) ([]EventChange, error)
	ListEvents(context.Context, EventListQuery) (Page[Event], error)
	ListEventsByOrganizer(context.Context, string) ([]Event, error)
	CreateRoadbook(context.Context, Roadbook) (Roadbook, error)
	GetRoadbook(context.Context, string) (*Roadbook, error)
	ListRoadbooks(context.Context, int, *string) (Page[Roadbook], error)
	RegisterAtomically(context.Context, RegisterCommand) (RegistrationResult, error)
	CancelRegistrationAtomically(context.Context, string, string, time.Time) (RegistrationResult, error)
	GetRegistration(context.Context, string, string) (*Registration, error)
	ListRegistrationsByUser(context.Context, string) ([]UserRegistration, error)
	ListEventParticipants(context.Context, string) ([]EventParticipant, error)
	GetEventParticipantContact(context.Context, string, string) (*EventParticipantContact, error)
}
